File: 0001-0099/0063-unique-paths-ii/0063-unique-paths-ii.cpp
#include <vector>

using namespace std;

class Solution {
public:
	int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
		// Get dimensions of the input
		int m = obstacleGrid.size();
		if (m == 0)
			return 0;
		int n = obstacleGrid[0].size();
		
		// Init the cell values to 0, add an extra row and column (the actual bottom row and
		// right-most column's values will be based on the extra row and column) to simplify the
		// calculations below.
		long mem[m + 1][n + 1] = {};
		for (int i = 0; i <= m; i++)
			for (int j = 0; j <= n; j++)
				mem[i][j] = 0;
		
		// Init base value in the extra row (or column, which is mem[m][n-1])
		mem[m - 1][n] = 1;
		
		for (int i = m - 1; i >= 0; i--)
			for (int j = n - 1; j >= 0; j--)
				// Check if cell is obstacle, mark as 0 if it is
				mem[i][j] = (obstacleGrid[i][j]) ? 0 : (mem[i + 1][j] + mem[i][j + 1]);
		
		return mem[0][0];
	}
};

File: 0001-0099/0063-unique-paths-ii/0063-unique-paths-ii.py
"""
Similar to #62 Unique Paths, just check for obstacle cells and skip counting paths that pass through them.
"""

class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        # If start and end cells are obstacles, there is 0 path
        if obstacleGrid[0][0] or obstacleGrid[-1][-1]:
            return 0

        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0] * n for _ in range(m)]
        dp[0][0] = 1

        for i in range(m):
            for j in range(n):
                # If current cell is obstable, skip
                if obstacleGrid[i][j]:
                    continue
                # Update method to reach cell below (no need to check obstacle, which is skipped when reaching tha
                # cell anyway)
                if i < m-1:
                    dp[i+1][j] += dp[i][j]
                # Update method to reach cell on the right
                if j < n-1:
                    dp[i][j+1] += dp[i][j]

        return dp[-1][-1]
